diff -ur '--exclude=.git' .unikraft.orig/libs/libsodium/Makefile.uk .unikraft/libs/libsodium/Makefile.uk
--- .unikraft.orig/libs/libsodium/Makefile.uk	2023-04-14 21:39:06.505000000 -0600
+++ .unikraft/libs/libsodium/Makefile.uk	2023-04-15 20:42:20.996000000 -0600
@@ -88,6 +88,7 @@
                      -DPACKAGE_URL=\"https://github.com/jedisct1/libsodium\" \
                      -DPACKAGE=\"libsodium\" \
                      -DVERSION=\"$(LIBSODIUM_VERSION)\" \
+                     -DSODIUM_STATIC=1 \
                      -DSODIUM_LIBRARY_MINIMAL=1 \
                      -D__linux__ \
                      -DSTDC_HEADERS=1 \
diff -ur '--exclude=.git' .unikraft.orig/libs/python3/Config.uk .unikraft/libs/python3/Config.uk
--- .unikraft.orig/libs/python3/Config.uk	2023-04-14 21:39:07.947000000 -0600
+++ .unikraft/libs/python3/Config.uk	2023-04-15 18:41:43.936000000 -0600
@@ -80,6 +80,13 @@
 	bool "zlib"
 	default y
 endif
+config LIBPYTHON3_EXTENSION_SODIUM
+    bool "libsodium"
+    default n
+    depends on LIBSODIUM
+config LIBPYTHON3_EXTENSION_SODIUM_MODULE_C
+	string "C file for _sodium Module"
+	default "_sodium.c"
 endif
 
 endif
diff -ur '--exclude=.git' .unikraft.orig/libs/python3/Makefile.uk .unikraft/libs/python3/Makefile.uk
--- .unikraft.orig/libs/python3/Makefile.uk	2023-04-14 21:39:07.947000000 -0600
+++ .unikraft/libs/python3/Makefile.uk	2023-04-15 20:58:56.844000000 -0600
@@ -367,7 +367,7 @@
 endif
 ifeq ($(CONFIG_LIBPYTHON3_EXTENSION_EXPAT),y)
 LIBPYTHON3_EXTENSIONS_CINCLUDES-y += -I$(LIBPYTHON3_SRC)/Modules/expat
-LIBPYTHON3_EXTENSIONS_CFLAGS-y    += -DHAVE_EXPAT_CONFIG_H=1 -DXML_POOR_ENTROPY=1 -DUSE_PYEXPAT_CAPI
+LIBPYTHON3_EXTENSIONS_CFLAGS-y    += -DHAVE_EXPAT_CONFIG_H=1 -DXML_POOR_ENTROPY=1 -DUSE_PYEXPAT_CAPI -DSODIUM_LIBRARY_MINIMAL=1
 LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Modules/_elementtree.c
 LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Modules/expat/loadlibrary.c
 LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Modules/expat/xmlparse.c
@@ -415,6 +415,7 @@
 LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Modules/_sha3/sha3module.c
 LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Modules/sha512module.c
 LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Modules/socketmodule.c
+LIBPYTHON3_EXTENSIONS_SRCS-$(CONFIG_LIBPYTHON3_EXTENSION_SODIUM) += $(LIBPYTHON3_SRC)/Modules/_sodiummodule.c
 # TODO enable LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Modules/spwdmodule.c
 ifeq ($(CONFIG_LIBPYTHON3_EXTENSION_SQLITE),y)
 LIBPYTHON3_EXTENSIONS_CACHE_FLAGS-$(CONFIG_LIBPYTHON3_EXTENSION_SQLITE) += -DMODULE_NAME='"sqlite3"' -DSQLITE_OMIT_LOAD_EXTENSION
@@ -455,6 +456,8 @@
 
 #LIBPYTHON3_EXTENSIONS_SRCS-y += $(LIBPYTHON3_SRC)/Misc/coverity_model.c
 
+LIBPYTHON3_EXTENSION_SODIUM_MODULE_C = $(shell echo $(CONFIG_LIBPYTHON3_EXTENSION_SODIUM_MODULE_C))
+
 ################################################################################
 # Prepare rules
 ################################################################################
@@ -465,6 +468,10 @@
 		cp $(APP_BASE)/.config $@ && \
 		sed -i 's/^CONFIG_STACK_SIZE_PAGE_ORDER=.*$$/CONFIG_STACK_SIZE_PAGE_ORDER=10/g' $(APP_BASE)/.config)
 
+# Copy constructed pynacl module (created outside normal processing)
+$(LIBPYTHON3_SRC)/Modules/_sodiummodule.c: $(BUILD_DIR)/$(LIBPYTHON3_EXTENSION_SODIUM_MODULE_C)
+	$(call verbose_cmd,CONFIG,libpython3: $(notdir $@), cp $< $@)
+
 LIBPYTHON3_PREPARED_DEPS = \
 	$(APP_BASE)/.config.orig \
 
diff -ur '--exclude=.git' .unikraft.orig/libs/python3/modules_config.c .unikraft/libs/python3/modules_config.c
--- .unikraft.orig/libs/python3/modules_config.c	2023-04-14 21:39:07.948000000 -0600
+++ .unikraft/libs/python3/modules_config.c	2023-04-15 14:26:59.676000000 -0600
@@ -119,6 +119,7 @@
 extern PyObject* PyInit__sha512(void);
 extern PyObject* PyInit__signal(void);
 extern PyObject* PyInit__socket(void);
+extern PyObject* PyInit__sodium(void);
 extern PyObject* PyInit_spwd(void);
 extern PyObject* PyInit__ssl(void);
 extern PyObject* PyInit__stat(void);
@@ -289,6 +290,9 @@
     {"_sha512", PyInit__sha512},
     {"_signal", PyInit__signal},
     {"_socket", PyInit__socket},
+#if CONFIG_LIBPYTHON3_EXTENSION_SODIUM
+    {"_sodium", PyInit__sodium},
+#endif
 //TODO enable    {"spwd", PyInit_spwd},
 #if CONFIG_LIBPYTHON3_EXTENSION_SQLITE
     {"sqlite3", PyInit__sqlite3},
